### PR TITLE
Fix compilation for Julia builds without GPL libraries

### PR DESCRIPTION
**Pull request by jonathan-conder-sm**
I'm trying to use `Flux.jl`, which works fine without `SuiteSparse` provided you don't use any sparse matrices. However it depends on this package, which fails to compile if `Base.USE_GPL_LIBS == false`.

### CODE REVIEW

File: src/SparseInverseSubset.jl [hunk 1]
```diff
@@ -42,15 +42,19 @@ end
...
-    P = F.p
-    d = Vector(diag(L))
-    L = tril(L * Diagonal(1 ./ d), -1)
-    U = sparse(L')
-    d = d.^2
-    D = Diagonal(d)
-    return (L=L, D=D, U=U, P=P)
+function get_ldup end
+
+if Base.USE_GPL_LIBS
```
Comment (ElOceanografo) [inline src/SparseInverseSubset.jl:47]: Should this conditional go outside the function definition?

Comment (jonathan-conder-sm) [reply src/SparseInverseSubset.jl:47]: up to you really. I did it this way so the function will still be documented even when `Base.USE_GPL_LIBS` is `false`

**Comment by ElOceanografo** [comment]
Can you confirm what the dependency chain is from Flux to here? I assume it's Zygote > ChainRules > SparseInverseSubset. (This package is required to calculate the rrule for the logdet of a sparse matrix, see https://github.com/JuliaDiff/ChainRules.jl/pull/730.) 

If that's the case, I think it might make more sense to put these conditionals in ChainRules? That package is still importing SuiteSparse and relying on its types, so I suspect you'd still get these compilation errors (though I haven't checked).

**Comment by jonathan-conder-sm** [comment]
> Can you confirm what the dependency chain is from Flux to here? I assume it's Zygote > ChainRules > SparseInverseSubset.

Yeah that's right

> If that's the case, I think it might make more sense to put these conditionals in ChainRules? That package is still importing SuiteSparse and relying on its types, so I suspect you'd still get these compilation errors (though I haven't checked).

It works fine after this patch. It's only CHOLMOD, SPQR and UMFPACK that are missing from SuiteSparse.

**Comment by ElOceanografo** [comment]
After some more consideration, I think this logic really does belong in ChainRules rather than here. `sparseinv(A::SparseMatrixCSC)` just takes the Cholesky decomposition and then passes that factorization to `sparseinv(F::SuiteSparse.CHOLMOD.Factor)`, which calls `get_ldup`, so the whole package will end up broken on non-GPL builds of Julia.

I think a better approach would be not defining the `rrule`s for (log)determinants of sparse matrices in ChainRules if `USE_GPL_LIBS == false`, or perhaps implement an alternative version of them that raises an informative error saying that they depend on GPL libraries and data structures and so aren't available on GPL-free builds.

**Comment by jonathan-conder-sm** [comment]
Makes sense to me.

I'm not going to do any more work on this though: after I made this PR I realised I can set `USE_GPL_LIBS = true` and just not distribute the GPL binaries. There's still an error message at compile time, from `SparseArrays.__init__`, but it doesn't prevent compilation and is easy to patch out.

**Pull request closed without merging**